### PR TITLE
artifact: git needs more files for private repositories

### PR DESCRIPTION
**Pull request by shoenig**
This PR fixes artifact downloading so that git may work when cloning from
private repositories. It needs

- file read on `/etc/passwd`
- dir read on `/root/.ssh`
- file write on `/root/.ssh/known_hosts`

Add these rules to the landlock rules for the artifact sandbox.

Fixes https://github.com/hashicorp/nomad/issues/16477

### CODE REVIEW

File: client/allocrunner/taskrunner/getter/util_linux.go [hunk 1]
```diff
@@ -42,12 +43,35 @@ func credentials() (uint32, uint32) {
...
+
+	// if home is set, just use that
+	if home, err := os.UserHomeDir(); err == nil {
+		return home
+	}
+
+	// if we are the root user, use "/root"
+	if unix.Getuid() == 0 {
+		return "/root"
+	}
```
Comment (tgross) [inline client/allocrunner/taskrunner/getter/util_linux.go:61]: We have `mitchell/go-homedir` in our dependencies already, which checks `getent` to get the homedir out of `/etc/passwd` if `$HOME` isn't set: https://github.com/mitchellh/go-homedir/blob/main/homedir.go#L114 Should we use that to handle weird edge cases like `$HOME` being unset for non-root users?

File: client/allocrunner/taskrunner/getter/util_linux.go [hunk 2]
```diff
@@ -71,26 +95,60 @@ func lockdown(allocDir, taskDir string) error {
...
 		_, err := os.Stat(p)
 		return err == nil
 	}
-	result := make([]*landlock.Path, 0, 3)
+
+	result := make([]*landlock.Path, 0, 6)
+	if exists(sshDir) {
+		result = append(result, landlock.Dir(sshDir, "r"))
+	}
+	if exists(knownHosts) {
```
Comment (tgross) [inline client/allocrunner/taskrunner/getter/util_linux.go:146]: Does landlock require that the file already exists? I'm thinking of the case where the user has just spun up a VM from scratch and `/root/.ssh/known_hosts` hasn't been created yet.

File: client/allocrunner/taskrunner/getter/util_linux.go [hunk 3]
```diff
@@ -42,12 +43,35 @@ func credentials() (uint32, uint32) {
...
+		return home
+	}
+
+	// if we are the root user, use "/root"
+	if unix.Getuid() == 0 {
+		return "/root"
+	}
+
+	// nothing safe to do
+	return "/dev/null"
```
Comment (schmichael) [inline client/allocrunner/taskrunner/getter/util_linux.go:64]:
```suggestion
	return "/nonexistent"
```

Let's use what `nobody` uses for home. I'd hate for an accidental misuse of this func (even in a test) to overwrite someone's `/dev/null` or similar. :sweat_smile:

File: client/allocrunner/taskrunner/getter/util_linux.go (hunk 2 above)
Comment (shoenig) [reply client/allocrunner/taskrunner/getter/util_linux.go:146]: > Does landlock require that the file already exists? 

Yup. But this code is running in the artifact downloader child process - if you get to this point without your ssh config, you're gonna have a bad time anyway.

File: client/allocrunner/taskrunner/getter/util_linux.go (hunk 1 above)
Comment (shoenig) [reply client/allocrunner/taskrunner/getter/util_linux.go]: Good call! Swapped out UserHomeDir for go-homedir; this makes the tests a bit more interesting to setup but seems worth it.

**Review by tgross — APPROVED** [review]
LGTM :+1:

**Pull request merged**